src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ParseJson_Coercions.txt

```diff
@@ -0,0 +1,320 @@
+﻿// The If function will attempt to coerce the else clause to the same type as the then clause
+// this allows us to control the exact expected coercion type
+>> If(false, 4, ParseJSON("5"))
+5
+
+>> If(true, 4, ParseJSON("5"))
+4
+
+>> If(false, "s", ParseJSON("""t"""))
+"t"
+
+>> If(ParseJSON("true"), 1, 0)
+1
+
+>> If(false, ColorValue("#aabbcc"), ParseJSON("""#aabbcc"""))
+RGBA(170,187,204,1)
+
+>> If(false, GUID("5cc45615-f759-4a53-b225-d3a2497f60ad"), ParseJSON("""5cc45615-f759-4a53-b225-d3a2497f60ad"""))
+GUID("5cc45615-f759-4a53-b225-d3a2497f60ad")
+
+>> DateDiff(If(false, DateTime(0,0,0,0,0,0), DateTimeValue(ParseJSON("""2011-01-15T08:00:00.000Z"""))), DateTimeValue(ParseJSON("""2011-01-30T08:00:00.000Z""")))
+15
+
+>> DateDiff(If(false, Date(0,0,0), DateValue(ParseJSON("""2011-01-15T00:00:00"""))), DateValue(ParseJSON("""2011-01-30""")))
+15
+
+>> Text(If(false, Time(0,0,0), TimeValue(ParseJSON("""08:03:05.000"""))), "HH:mm:ss")
+"08:03:05"
+
+>> Sum(ParseJSON("[1,2,3]"), Blank())
+Errors: Error 4-24: Untyped objects cannot be used as the first argument to functions which support record scopes.|Error 0-34: The function 'Sum' has some invalid arguments.
+
+// Arity error code path for Sum
+>> Sum(ParseJSON("[1,2,3]"))
+Errors: Error 4-24: Untyped objects cannot be used as the first argument to functions which support record scopes.|Error 0-25: The function 'Sum' has some invalid arguments.
+
+// Scalar version of Sum
+>> Sum([ParseJSON("1"), ParseJSON("2"), ParseJSON("3")], Value)
+6
+
+>> Filter(ParseJSON("[1,2,3]"), true)
+Errors: Error 7-27: Untyped objects cannot be used as the first argument to functions which support record scopes.|Error 0-34: The function 'Filter' has some invalid arguments.
+
+// Comparison tests
+>> Abs(5) = Abs(ParseJSON("5"))
+true
+
+>> Acot(5) = Acot(ParseJSON("5"))
```

Comment:
Are there any invariants we can establish here? 

eg, like: 

Value(str) == Value(ParseJson(str)) ? 

DateValue(str) == DateValue(ParseJson(QUOTE & str & QUOTE))

etc?

Reply:
Coercions are disabled on constructor functions, so that would be unrelated to this PR. Beyond that, we don't guarantee these invariants. Some constructor functions have much stricter format requirements to encourage better API design.